agents/agents-features/agents-features-opentelemetry/src/jvmTest/kotlin/ai/koog/agents/features/opentelemetry/langfuse/TracingTest.kt

```diff
@@ -0,0 +1,498 @@
+     *                     This is necessary because the tests use TestGetWeatherTool, which needs to be registered.
+     */
+    private suspend fun runAgentWithStrategy(
+        strategy: AIAgentStrategy<String, String>,
+        userPrompt: String = "User prompt message",
+        mockExecutor: PromptExecutor = getMockExecutor {
+            mockLLMAnswer("The weather in Paris is rainy and overcast, with temperatures around 57°F") onRequestEquals userPrompt
+        },
+        toolRegistry: ToolRegistry = ToolRegistry {
+            tool(TestGetWeatherTool)
+        },
+    ) {
+        val agentId = "test-agent-id"
+        val promptId = "test-prompt-id"
+        val testClock = Clock.System
+        val model = OpenAIModels.Chat.GPT4o
+        val temperature = 0.4
+
+        val agent = createAgent(
+            agentId = agentId,
+            strategy = strategy,
+            promptId = promptId,
+            promptExecutor = mockExecutor,
+            model = model,
+            clock = testClock,
+            temperature = temperature,
+            toolRegistry = toolRegistry,
+        ) {
+            install(OpenTelemetry) {
+                addSpanExporter(
+                    createLangfuseSpanExporter(
+                        System.getenv()["LANGFUSE_HOST"]
+                            ?: throw IllegalArgumentException("LANGFUSE_HOST is not set"),
+                        System.getenv()["LANGFUSE_PUBLIC_KEY"]
+                            ?: throw IllegalArgumentException("LANGFUSE_PUBLIC_KEY is not set"),
+                        System.getenv()["LANGFUSE_SECRET_KEY"]
+                            ?: throw IllegalArgumentException("LANGFUSE_SECRET_KEY is not set"),
+                    )
+                )
+                addSpanExporter(
+                    createWeaveSpanExporter(
+                        "https://trace.wandb.ai",
+                        System.getenv()["WEAVE_ENTITY"] ?: throw IllegalArgumentException("WEAVE_ENTITY is not set"),
+                        System.getenv()["WEAVE_PROJECT_NAME"] ?: "koog-tracing",
+                        System.getenv()["WEAVE_API_KEY"] ?: throw IllegalArgumentException("WEAVE_API_KEY is not set")
+                    )
+                )
+                addSpanExporter(inMemorySpanExporter)
+            }
+        }
+
+        agent.use {
+            it.run(userPrompt)
+        }
+    }
+}
+
+private fun createLangfuseSpanExporter(
+    langfuseUrl: String,
+    langfusePublicKey: String,
+    langfuseSecretKey: String,
+): SpanExporter {
+    val credentials = "$langfusePublicKey:$langfuseSecretKey"
+    val auth = Base64.getEncoder().encodeToString(credentials.toByteArray(Charsets.UTF_8))
+
+    return OtlpHttpSpanExporter.builder()
+        .setTimeout(30, TimeUnit.SECONDS)
+        .setEndpoint("$langfuseUrl/api/public/otel/v1/traces")
+        .addHeader("Authorization", "Basic $auth")
+        .build()
+}
+
+private fun createWeaveSpanExporter(
```

Comment:
major: It seems for me that test code should not include a production code that creates a real client. Here and above, see `createLangfuseSpanExporter`.